File: src/query.ts
import { FileSummary, ProjectIndex } from "./storage.js";

// ─── Query Engine (100% local, zero API cost) ─────────────────────────────────

export interface QueryResult {
  file: FileSummary;
  score: number;
  matchedOn: string[];
}

export function queryIndex(
  index: ProjectIndex,
  userQuery: string,
  topK: number = 10
): QueryResult[] {
  const queryTokens = tokenize(userQuery);
  const results: QueryResult[] = [];

  for (const [, file] of Object.entries(index.files)) {
    const { score, matchedOn } = scoreFile(file, queryTokens);
    if (score > 0) {
      results.push({ file, score, matchedOn });
    }
  }

  // Sort by score descending, return top K
  return results
    .sort((a, b) => b.score - a.score)
    .slice(0, topK);
}

// ─── Scoring ──────────────────────────────────────────────────────────────────

function scoreFile(
  file: FileSummary,
  queryTokens: string[]
): { score: number; matchedOn: string[] } {
  let score = 0;
  const matchedOn: string[] = [];

  const summaryTokens = tokenize(file.summary);
  const tagTokens = file.tags.map((t) => t.toLowerCase());
  const exportTokens = file.exports.map((e) => e.toLowerCase());
  const depTokens = file.dependencies.map((d) => d.toLowerCase());
  const pathTokens = tokenize(file.path);

  for (const qt of queryTokens) {
    // File path match — high weight (file name is very relevant)
    if (pathTokens.some((pt) => pt.includes(qt) || qt.includes(pt))) {
      score += 3;
      matchedOn.push(`path:${qt}`);
    }

    // Tag match — high weight (tags are curated keywords)
    if (tagTokens.some((tag) => tag.includes(qt) || qt.includes(tag))) {
      score += 2.5;
      matchedOn.push(`tag:${qt}`);
    }

    // Summary match — medium weight
    if (summaryTokens.some((st) => st.includes(qt) || qt.includes(st))) {
      score += 2;
      matchedOn.push(`summary:${qt}`);
    }

    // Export match — medium weight
    if (exportTokens.some((et) => et.includes(qt) || qt.includes(et))) {
      score += 1.5;
      matchedOn.push(`export:${qt}`);
    }

    // Dependency match — lower weight
    if (depTokens.some((dt) => dt.includes(qt) || qt.includes(dt))) {
      score += 1;
      matchedOn.push(`dep:${qt}`);
    }
  }

  // Deduplicate matchedOn
  const unique = [...new Set(matchedOn)];

  return { score, matchedOn: unique };
}

// ─── Tokenizer ────────────────────────────────────────────────────────────────

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s\-_\/\.]/g, " ")
    .split(/[\s\-_\/\.]+/)
    .filter((t) => t.length > 2)
    .filter((t) => !STOP_WORDS.has(t));
}

const STOP_WORDS = new Set([
  "the", "and", "for", "with", "this", "that", "from", "into",
  "are", "was", "were", "has", "have", "had", "not", "but",
  "its", "it", "is", "in", "of", "to", "a", "an", "or",
  "can", "will", "should", "would", "could", "may", "might",
]);

File: python/smartctx/query.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import List

from smartctx.storage import FileSummary, ProjectIndex

# ── Query Engine (100% local, zero API cost) ─────────────────────────────────


@dataclass
class QueryResult:
    file: FileSummary
    score: float
    matchedOn: List[str]


def query_index(index: ProjectIndex, user_query: str, top_k: int = 10) -> List[QueryResult]:
    query_tokens = _tokenize(user_query)
    results: List[QueryResult] = []

    for file in index.files.values():
        score, matched_on = _score_file(file, query_tokens)
        if score > 0:
            results.append(QueryResult(file=file, score=score, matchedOn=matched_on))

    results.sort(key=lambda r: r.score, reverse=True)
    return results[:top_k]


# ── Scoring ──────────────────────────────────────────────────────────────────

STOP_WORDS = {
    "the", "and", "for", "with", "this", "that", "from", "into",
    "are", "was", "were", "has", "have", "had", "not", "but",
    "its", "it", "is", "in", "of", "to", "a", "an", "or",
    "can", "will", "should", "would", "could", "may", "might",
}


def _tokenize(text: str) -> List[str]:
    lowered = text.lower()
    cleaned = re.sub(r"[^a-z0-9\s\-_/.]", " ", lowered)
    tokens = re.split(r"[\s\-_/.]+", cleaned)
    return [t for t in tokens if len(t) > 2 and t not in STOP_WORDS]


def _score_file(file: FileSummary, query_tokens: List[str]) -> tuple:
    score = 0.0
    matched_on: List[str] = []

    summary_tokens = _tokenize(file.summary)
    tag_tokens = [t.lower() for t in file.tags]
    export_tokens = [e.lower() for e in file.exports]
    dep_tokens = [d.lower() for d in file.dependencies]
    path_tokens = _tokenize(file.path)

    for qt in query_tokens:
        # File path match — high weight
        if any(qt in pt or pt in qt for pt in path_tokens):
            score += 3
            matched_on.append(f"path:{qt}")

        # Tag match — high weight
        if any(qt in tag or tag in qt for tag in tag_tokens):
            score += 2.5
            matched_on.append(f"tag:{qt}")

        # Summary match — medium weight
        if any(qt in st or st in qt for st in summary_tokens):
            score += 2
            matched_on.append(f"summary:{qt}")

        # Export match — medium weight
        if any(qt in et or et in qt for et in export_tokens):
            score += 1.5
            matched_on.append(f"export:{qt}")

        # Dependency match — lower weight
        if any(qt in dt or dt in qt for dt in dep_tokens):
            score += 1
            matched_on.append(f"dep:{qt}")

    return score, list(set(matched_on))
